src/coupling/assembly_hm.hh

```diff
@@ -0,0 +1,190 @@
+                    // i.e. where only mechanical traction is prescribed.
+                    auto p_bdr = p.point_bdr(dh_side.cond().element_accessor());
+                    double alpha = eq_fields_->alpha(p);
+                    double density = eq_fields_->density(p);
+                    double gravity = eq_fields_->gravity(p);
+                    double bc_pressure = eq_data_->flow_->eq_fields().bc_pressure(p_bdr);
+                    ref_pot += -alpha*density*gravity*bc_pressure * fe_values_side_.JxW(k) / dh_side.measure();
+                    ++k;
+                }
+            }
+        }
+        ref_potential_vec_.set(dof_indices_[dh_side.side_idx()], ref_pot);
+    }
+
+
+
+private:
+
+    EqFields *eq_fields_;       ///< Fields shared with HM_Iterative
+    EqData *eq_data_;           ///< Data objects shared with HM_Iterative
+
+    /// Sub field set contains fields used in calculation.
+    FieldSet used_fields_;
+
+    FEValues<3> fe_values_side_;                              ///< FEValues of side object
+    LocDofVec dof_indices_;                             ///< Vector of global DOF indices
+    VectorMPI ref_potential_vec_;                             ///< Vector of dofs of field ref_potential
+
+    template < template<IntDim...> class DimAssembly>
+    friend class GenericAssembly;
+
+};
+
+
+template <unsigned int dim>
+class ResidualAssemblyHM : public AssemblyBase<dim>
+{
+public:
+    typedef typename HM_Iterative::EqFields EqFields;
+    typedef typename HM_Iterative::EqData EqData;
+
+    static constexpr const char * name() { return "ResidualAssemblyHM"; }
+
+    /// Constructor.
+    ResidualAssemblyHM(EqFields *eq_fields, EqData *eq_data)
+    : AssemblyBase<dim>(1), eq_fields_(eq_fields), eq_data_(eq_data) {
+        this->active_integrals_ = (ActiveIntegrals::bulk);
+        this->used_fields_ += eq_data_->flow_->eq_fields().field_ele_pressure;
+        this->used_fields_ += eq_fields_->old_iter_pressure;
+    }
+
+    /// Destructor.
+    ~ResidualAssemblyHM() {}
+
+    /// Initialize auxiliary vectors and other data members
+    void initialize(ElementCacheMap *element_cache_map) {
+        this->element_cache_map_ = element_cache_map;
+        shared_ptr<FE_P<dim>> fe_ = std::make_shared< FE_P<dim> >(0);
+        fe_values_.initialize(*this->quad_, *fe_, update_JxW_values);
+    }
+
+
+    /// Assemble integral over element
+    inline void cell_integral(DHCellAccessor cell, unsigned int element_patch_idx)
+    {
+        if (cell.dim() != dim) return;
+        if (!cell.is_own()) return;
+
+        fe_values_.reinit(cell.elm());
+
+        // compute pressure error
+        unsigned int k=0;
+        for (auto p : this->bulk_points(element_patch_idx) )
+        {
+            double new_p = eq_data_->flow_->eq_fields().field_ele_pressure(p);
+            double old_p = eq_fields_->old_iter_pressure(p);
+            eq_data_->p_dif2 += pow(new_p - old_p, 2)*fe_values_.JxW(k);
```

Comment:
Rather use multiplication for small integer powers, 10x times faster:
https://stackoverflow.com/questions/2940367/what-is-more-efficient-using-pow-to-square-or-just-multiply-it-with-itself.